### ISSUE TITLE
MetalFX upscaling on device to improve visual clarity/latency.

### ISSUE THREAD

**Issue by panthurrr** [alvr-org/alvr-visionos#36]
MetalFX has two upscaling types for visionOS which may improve the overall clarity and latency compared to native. SpatialScaler requires less data to configure and should let us know if temporal scaling is worth adding.

https://developer.apple.com/documentation/metalfx/

**Comment by xuhao1** [comment]
Yes I have tried this, see https://github.com/alvr-org/alvr-visionos/pull/97
However, due to the bandwidth limitation, visual effects does not change a lot.

**Comment by panthurrr** [comment]
It might be necessary to apply MSAA prior to upscaling for more noticeable benefits

I started to add this based on apple's apple's sample code:
https://developer.apple.com/documentation/metal/metal_sample_code_library/improving_edge-rendering_quality_with_multisample_antialiasing_msaa

had some challenges with the resolve textures matching at the time though this was a long time ago now.

Will look into this more now that you've got upscaling added, thanks for the addition!
